CREATING_TOXICS.md

```diff
@@ -0,0 +1,137 @@
+# Creating custom toxics
+
+Creating a toxic is done by implementing the `Toxic` interface:
+
+```go
+type Toxic interface {
+    Pipe(*toxics.ToxicStub)
+}
+```
+
+The `Pipe()` function defines how data flows through the toxic, and is passed a
+`ToxicStub` to operate on. A `ToxicStub` stores the input and output channels for
+the toxic, as well as an interrupt channel that is used to pause operation of the
+toxic.
+
+The input and output channels in a `ToxicStub` send and receive `StreamChunk` structs,
+which are similar to network packets. A `StreamChunk` contains a `byte[]` of stream
+data, and a timestamp of when Toxiproxy received the data from the client or server.
+
+Toxics are registered in an `init()` function so that they can be used by the server:
+```go
+func init() {
+    toxics.Register("toxic_name", new(ExampleToxic))
+}
+```
+
+## A basic toxic
+
+The most basic implementation of a toxic is the [noop toxic](https://github.com/Shopify/toxiproxy/blob/master/toxics/noop.go),
+which just passes data through without any modifications.
+
+```go
+type NoopToxic struct{}
+
+func (t *NoopToxic) Pipe(stub *toxics.ToxicStub) {
+    for {
+        select {
+        case <-stub.Interrupt:
+            return
+        case c := <-stub.Input:
+            if c == nil {
+                stub.Close()
+                return
+            }
+            stub.Output <- c
+        }
+    }
+}
+```
+
+The above code reads from `stub.Input` in a loop, and passes the `StreamChunk` along to
+`stub.Output`. Since reading from `stub.Input` will block until a chunk is available,
+we need to check for interrupts as the same time.
+
+Toxics will be interrupted whenever they are being updated, or possibly removed. This can
+happen at any point within the `Pipe()` function, so all blocking operations (including sleep),
+should be interruptible. When an interrupt is received, the toxic should return from the `Pipe()`
+function after it has written any "in-flight" data back to `stub.Output`. It is important that
+all data read from `stub.Input` is passed along to `stub.Output`, otherwise the stream will be
+missing bytes and become corrupted.
+
+When an `end of stream` is reached, `stub.Input` will begin returning `nil` chunks. Whenever a
```

Comment:
When you `close()` a channel, all further reads will return nil. The toxic should never read more than one `nil` chunk, so I'll just change the wording here.